### ISSUE TITLE
Test not reported as failed when class spring wiring fails

### ISSUE THREAD

**Issue by jordanbragg** [serenity-bdd/serenity-core#352]
I am currently having issues where if something being injected or the configuration is wrong in ANY way, the test silently throws an error and continues on excluding the test from the overall results. This is a pretty large problem for our team. I will try to look into it as well, but I have an example at:

https://github.com/jordanbragg/serenity-report-test/blob/master/src/test/java/reporting/spring/FailedSpringWiring.java

I made it fail for obvious reasons here, but a real use case I am having is injecting a standalone JPA that points to a database that may not be reachable or up. If it fails during integration testing in Jenkins, we will never know as it reports success.

Here is the output: 

gradle cleanTest test
:compileJava UP-TO-DATE
:processResources UP-TO-DATE
:classes UP-TO-DATE
:compileTestJava
:processTestResources UP-TO-DATE
:testClasses
:test

r14:57:13.558 [DEBUG] [TestEventLogger] Gradle Test Executor 1 STARTED
14:57:13.562 [QUIET] [system.out] 14:57:13.562 [DEBUG] [org.gradle.api.internal.tasks.testing.junit.JUnitTestClassProcessor] Executing test class reporting.spring.FailedSpringWiring
14:57:13.563 [DEBUG] [TestEventLogger] 
14:57:13.563 [DEBUG] [TestEventLogger] reporting.spring.FailedSpringWiring STARTED
14:57:13.830 [DEBUG] [TestEventLogger] 
14:57:13.830 [DEBUG] [TestEventLogger] reporting.spring.FailedSpringWiring STANDARD_ERROR
14:57:13.830 [DEBUG] [TestEventLogger]     SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
14:57:13.830 [DEBUG] [TestEventLogger]     SLF4J: Defaulting to no-operation (NOP) logger implementation
14:57:13.830 [DEBUG] [TestEventLogger]     SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
14:57:13.988 [DEBUG] [TestEventLogger]     Mar 10, 2016 2:57:13 PM org.springframework.test.context.support.AbstractContextLoader generateDefaultLocations
14:57:13.988 [DEBUG] [TestEventLogger]     INFO: Could not detect default resource locations for test class [reporting.spring.FailedSpringWiring]: no resource found for suffixes {-context.xml, Context.groovy}.
14:57:13.989 [DEBUG] [TestEventLogger]     Mar 10, 2016 2:57:13 PM org.springframework.test.context.support.AnnotationConfigContextLoaderUtils detectDefaultConfigurationClasses
14:57:13.989 [DEBUG] [TestEventLogger]     INFO: Could not detect default configuration classes for test class [reporting.spring.FailedSpringWiring]: FailedSpringWiring does not declare any static, non-private, non-final, inner classes annotated with @Configuration.
14:57:14.110 [QUIET] [system.out] 14:57:14.110 [DEBUG] [org.gradle.api.internal.tasks.testing.junit.JUnitTestClassProcessor] Executing test class reporting.test.TestReportWithParameterTest
14:57:14.111 [DEBUG] [TestEventLogger] 

BUILD SUCCESSFUL

**Comment by jordanbragg** [comment]
Hey @YamStranger ,

So I finally got some time to look into this more and it appears to be related to the retry test functionality. Basically the Spring tests are failing in the ParentRunner.classBlock method executing withClassRules. What happens though is ParentRunner handled the exception and fires a testFailed before runChild ever executes and RetryFilteringRunNotifier never triggers a failure on the target RunNotifier that is passed in. 

Whats odd is that RetryFilteringRunNotifier appears to be the notifier to use when doing retries, but it only gets used when junit.retry.tests is false or not present. If I set junit.retry.tests to true, the tests properly show up with failures because the RunNotifier is not being wrapped by the RetryFilteringRunNotifier.

I think I have a fix, which is basically first to fix the default to not automatically wrap the RunNotifier in RetryFilteringRunNotifier unless junit.retry.tests is true, which adds a bunch of complexity if I don't care about retries. The second fix would be in RetryFilteringRunNotifier.fireTestFailure where, if the testStartAlreadyFired is not yet true, fireTestFailure on the target listener as well. Does this make sense?

A question I have too, in the test WhenRetryingFailedTests, the assertThat(notifier.failed, is(true)); assertion, why was that switched from false to true recently? If you try the test and the second time it succeeds, why would you want the build to fail?

Thanks!

**Comment by YamStranger** [comment]
Hello @jordanbragg 

Thanks for your proposal about fix will take a look.

>A question I have too, in the test WhenRetryingFailedTests, the assertThat(notifier.failed, is(true)); assertion, why was that switched from false to true recently? If you try the test and the second time it succeeds, why would you want the build to fail?

It was fix during pull request: https://github.com/serenity-bdd/serenity-core/pull/171/files

Lines below check that even if test failed during one retry, test should be success if last attempt allowed by retries was success. But based on custom notifier implementation - it will keep failed flag even if it will be notified about success at the and, based on retry mechanism:  
```
assertThat(outcomes.size(), is(1));
assertThat(outcomes.get(0).getResult(), is(TestResult.SUCCESS));
assertThat(notifier.failed, is(true));
```
If you take a look at used run notifier [implementation](https://github.com/serenity-bdd/serenity-core/blob/56b027874e67ebb0bda54959356a2725f5ec91c8/serenity-junit/src/test/java/net/serenitybdd/junit/runners/integration/WhenRetryingFailedTests.java#L85) you will find that if test was reported as fail - notifier will recorder it and this will **not changed** during second success run of test with same name. 

Internally notifier is wrapped with [retry-notifier](https://github.com/serenity-bdd/serenity-core/blob/56b027874e67ebb0bda54959356a2725f5ec91c8/serenity-junit/src/main/java/net/serenitybdd/junit/runners/RetryFilteringRunNotifier.java#L97), that can be [reseated ](https://github.com/serenity-bdd/serenity-core/blob/56b027874e67ebb0bda54959356a2725f5ec91c8/serenity-junit/src/main/java/net/serenitybdd/junit/runners/SerenityRunner.java#L439) if retry is enabled, and "wrapped notifier" will be notified about all fails and all success of same test during retry-enabled execution - and only "last" result should be keep and included in report.

**Comment by wakaleo** [comment]
Hi @jordanbragg, 

I think this issue should be fixed in version 1.1.32-rc.2, could you confirm?

**Comment by wakaleo** [comment]
Fixed in 1.1.33
